src/Elm.elm

```diff
@@ -1156,6 +1156,135 @@ lambda argBaseName argType toExpression =
+    in
+    case e of
+        Exp.LambdaExpression { args, expression } ->
+            case popLast args of
+                Just ( initLambdaArgs, Pattern.VarPattern lastLambdaArg ) ->
+                    case Compiler.denode expression of
+                        Exp.RecordAccess argNode fieldNode ->
+                            let
+                                fieldName =
+                                    Compiler.denode fieldNode
+
+                                arg =
+                                    Compiler.denode argNode
+                            in
+                            case arg of
+                                Exp.FunctionOrValue [] argName ->
+                                    if argName == lastLambdaArg then
+                                        Exp.RecordAccessFunction <| "." ++ fieldName
+
+                                    else
+                                        e
+
+                                _ ->
+                                    e
+
+                        Exp.Application applicationArgs ->
+                            case popLast applicationArgs of
+                                Just ( [], uniqueApplicationArg ) ->
+                                    if extractLastArg uniqueApplicationArg == Just lastLambdaArg then
+                                        Exp.FunctionOrValue [] "identity"
+
+                                    else
+                                        e
+
+                                Just ( initApplicationArgs, lastApplicationArg ) ->
+                                    if extractLastArg lastApplicationArg == Just lastLambdaArg then
+                                        if List.isEmpty initLambdaArgs then
+                                            case initApplicationArgs of
+                                                [ s ] ->
+                                                    betaReduce <| Compiler.denode s
+
+                                                _ ->
+                                                    Exp.Application initApplicationArgs
+
+                                        else
+                                            betaReduce <|
+                                                Exp.LambdaExpression
+                                                    { args = initLambdaArgs
+                                                    , expression =
+                                                        Compiler.nodify <|
+                                                            Exp.Application initApplicationArgs
+                                                    }
+
+                                    else
+                                        e
+
+                                _ ->
+                                    e
+
+                        _ ->
+                            e
+
+                _ ->
+                    e
+
+        _ ->
+            e
```

Comment:
I really don't like this part, but I fear that squashing it would require being more eager in what we evaluate, which is costly